test/solution-template-validation-tests/test/mainTemplateTests.js

```diff
@@ -115,16 +117,17 @@ describe('mainTemplate.json file - ', () => {
             });
         });
 
-        it('apiVersions must NOT be retrieved using providers().apiVersions[n].  This function is non-deterministic', () => {
+        // providers().apiVersions[n] MUST not be present
+        describe('apiVersions must NOT be retrieved using providers().apiVersions[n].  This function is non-deterministic', () => {
             templateFileJSONObjects.forEach(templateJSONObject => {
                 var templateObject = templateJSONObject.value;
-                templateObject.should.have.property('resources');
                 var message = 'in file:' + templateJSONObject.filename + ' should NOT have api version determined by providers().';
-                Object.keys(templateObject.resources).forEach(resource => {
-                    var val = templateObject.resources[resource];
-                    if (val.apiVersion) {
-                        val.apiVersion.should.withMessage(getErrorMessage(val, templateJSONObject.filepath, message)).not.contain('providers(');
-                    }
+                var properties = Object.keys(templateObject);
+                var resources = templateObject.resources;
+                it.each(properties, "providers().apiVersions[n] must NOT be present in the template", function(element, next) {
+                    var val = JSON.stringify(templateObject[element]);
+                    val.should.withMessage('file:' + templateJSONObject.filename + ' property:' + element).not.match(/providers\([a-zA-Z0-9_., '\\-]*\)\.apiVersions\[0\]/);
```

Comment:
The regex should also catch '(' and ')' per our discussion yesterday

Reply:
Our regex now can catch '(' and ')', remind me if there is something else?